### ISSUE TITLE
Dynamic upstreams per hostname

### ISSUE THREAD

**Issue by hermandavid** [caddyserver/caddy#4645]
I'm testing the `dynamic` feature in `reverse_proxy` and I have the following `Caddyfile`.

```
:8081 {
	reverse_proxy {
		dynamic srv "{http.request.host}" {
			resolvers 127.0.0.1:5353
		}
	}
}
```

Now, the requests are cached for 1 minute by default, but the DNS query is not sent for each different `Host` header value. I have implemented a simple DNS server with custom logic and I would like to use this approach to dynamically route requests to different upstreams.

Is there any way I can have a separate cache space for each `Host` value or any other approach to achieve this?

**Comment by mholt** [comment]
Ah, it's because the cache key uses the address before expanding placeholders. Will push a hopeful fix in a minute.

**Comment by mholt** [comment]
@hermandavid Can you try a build at dc4d147388547515f77447d594024386b732e7d4 and see if that resolves the issue for you? I think it will, since we now use the expanded value for the cache keys.

**Comment by hermandavid** [comment]
@mholt I have tried the fix and now caddy won't even send the DNS query and returns 502 straight away.

```
2022/03/21 11:58:25.005	ERROR	http.handlers.reverse_proxy	failed getting dynamic upstreams; falling back to static upstreams	{"error": "lookup : no such host"}
2022/03/21 11:58:25.005	ERROR	http.log.error	no upstreams available	{"request": {"remote_ip": "127.0.0.1", "remote_port": "54590", "proto": "HTTP/1.1", "method": "GET", "host": "testxx.app.internal", "uri": "/", "headers": {"User-Agent": ["curl/7.82.0"], "Accept": ["*/*"]}}, "duration": 0.00008768, "status": 502, "err_id": "cyp2thbkz", "err_trace": "reverseproxy.statusError (reverseproxy.go:1160)"}
```

I'm using the same Caddyfile and switching back to previous build works as initially described in the issue.

**Comment by francislavoie** [comment]
Please turn on the `debug` global option. Does your request actually specify a `Host`?

**Comment by hermandavid** [comment]
I'm testing it with the following command. Also, caddy seems to recognize the `Host` since it prints it out in the logs.

```
curl -H "Host: testxx.app.internal" 127.0.0.1:8081
```

Here is the log with `debug` option enabled.

```
2022/03/21 12:33:37.941	DEBUG	http.reverse_proxy.upstreams.srv	refreshing SRV upstreams	{"service": "", "proto": "", "name": ""}
2022/03/21 12:33:37.941	ERROR	http.handlers.reverse_proxy	failed getting dynamic upstreams; falling back to static upstreams	{"error": "lookup : no such host"}
2022/03/21 12:33:37.941	ERROR	http.log.error	no upstreams available	{"request": {"remote_ip": "127.0.0.1", "remote_port": "54632", "proto": "HTTP/1.1", "method": "GET", "host": "testxx.app.internal", "uri": "/", "headers": {"User-Agent": ["curl/7.82.0"], "Accept": ["*/*"]}}, "duration": 0.000131207, "status": 502, "err_id": "fnbzm2v03", "err_trace": "reverseproxy.statusError (reverseproxy.go:1160)"}
```

I'm getting the same output in the logs even ehwn I set the `name` property statically.

```
:8081 {
	reverse_proxy {
		dynamic srv "testxx.app.internal" {
			resolvers 127.0.0.1:5353
		}
	}
}
```

Resolving the domain name manually works normally.


```
dig @127.0.0.1 -p 5353 testxx.app.internal SRV +tcp

; <<>> DiG 9.18.1 <<>> @127.0.0.1 -p 5353 testxx.app.internal SRV +tcp
; (1 server found)
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: NOERROR, id: 562
;; flags: qr rd; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0
;; WARNING: recursion requested but not available

;; QUESTION SECTION:
;testxx.app.internal.		IN	SRV

;; ANSWER SECTION:
testxx.app.internal.	3600	IN	SRV	10 10 443 eu.app.internal.

;; Query time: 1240 msec
;; SERVER: 127.0.0.1#5353(127.0.0.1) (TCP)
;; WHEN: Mon Mar 21 13:38:59 CET 2022
;; MSG SIZE  rcvd: 92
```
